plugins/pulp_rpm/plugins/importers/yum/existing.py

```diff
@@ -135,14 +135,15 @@ def check_all_and_associate(wanted, conduit, config, download_deferred, catalog)
                 # package file does not exist and downloading is not deferred.
                 if not download_deferred and not file_exists:
                     continue
-            catalog.add(unit, wanted[unit.unit_key_as_named_tuple].download_path)
-            if rpm_parse.signature_enabled(config):
-                try:
-                    rpm_parse.filter_signature(unit, config)
-                except PulpCodedException as e:
-                    _LOGGER.debug(e)
-                    continue
-            repo_controller.associate_single_unit(conduit.repo, unit)
+            if not unit.is_associated(conduit.repo.repo_id):
```

Comment:
Rather than do a unique repo membership query for each unit, did you consider retrieving all of the unit IDs in the repo at once, storing them in a ``set()``, and then just checking for membership in that set here? It might be quite a bit faster, since it eliminates a ton of DB round-trips. I think a collection of unit IDs shouldn't be too expensive memory-wise.

I would probably use ``repo_controller.get_associated_unit_ids()``

I think this would be easy to throw together and do a performance comparison.